hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/dao/BaseTransactionProcessor.java

```diff
@@ -726,9 +739,81 @@ private void prepareThenExecuteTransactionWriteOperations(
 		}
 	}
 
+	@Nullable
+	protected RequestPartitionId determineRequestPartitionIdForWriteEntries(
+			RequestDetails theRequestDetails, List<IBase> theEntries) {
+		if (!myPartitionSettings.isPartitioningEnabled()) {
+			return RequestPartitionId.allPartitions();
+		}
+
+		RequestPartitionId retVal = null;
+
+		for (var nextEntry : theEntries) {
+			RequestPartitionId nextRequestPartitionId = null;
+			String verb = myVersionAdapter.getEntryRequestVerb(myContext, nextEntry);
+			if (isNotBlank(verb)) {
+				BundleEntryTransactionMethodEnum verbEnum = BundleEntryTransactionMethodEnum.valueOf(verb);
+				switch (verbEnum) {
+					case GET:
+						continue;
+					case DELETE: {
+						String requestUrl = myVersionAdapter.getEntryRequestUrl(nextEntry);
+						if (isNotBlank(requestUrl)) {
+							IdType id = new IdType(requestUrl);
+							String resourceType = id.getResourceType();
+							ReadPartitionIdRequestDetails details =
+									ReadPartitionIdRequestDetails.forDelete(resourceType, id);
+							nextRequestPartitionId = myRequestPartitionHelperService.determineReadPartitionForRequest(
+									theRequestDetails, details);
+						}
+						break;
+					}
+					case PATCH: {
+						String requestUrl = myVersionAdapter.getEntryRequestUrl(nextEntry);
+						if (isNotBlank(requestUrl)) {
+							IdType id = new IdType(requestUrl);
+							String resourceType = id.getResourceType();
+							ReadPartitionIdRequestDetails details =
+									ReadPartitionIdRequestDetails.forPatch(resourceType, id);
+							nextRequestPartitionId = myRequestPartitionHelperService.determineReadPartitionForRequest(
+									theRequestDetails, details);
+						}
+						break;
+					}
+					case POST:
+					case PUT: {
+						IBaseResource resource = myVersionAdapter.getResource(nextEntry);
+						if (resource != null) {
+							String resourceType = myContext.getResourceType(resource);
+							nextRequestPartitionId = myRequestPartitionHelperService.determineCreatePartitionForRequest(
+								theRequestDetails, resource, resourceType);
+						}
+					}
+				}
+			}
+
+			if (nextRequestPartitionId == null) {
+				// continue
+			} else if (retVal == null) {
+				retVal = nextRequestPartitionId;
+			} else if (!retVal.equals(nextRequestPartitionId)) {
+				if (myHapiTransactionService.isRequiresNewTransactionWhenChangingPartitions()) {
+					String msg = myContext
+						.getLocalizer()
+						.getMessage(BaseTransactionProcessor.class, "multiplePartitionAccesses", theEntries.size());
+					throw new InvalidRequestException(Msg.code(2541) + msg);
+				} else {
+					retVal = retVal.mergeIds(nextRequestPartitionId);
+				}
+			}
```

Comment:
problem: I think this will fail with MS if we have multiple partitions assigned to the same db.  How about something like:
```suggestion
			} else if (!retVal.equals(nextRequestPartitionId)) {
				if (!myHapiTransactionService.areCompatiblePartitions(retVal, nextRequestPartitionId)) {
					String msg = myContext
						.getLocalizer()
						.getMessage(BaseTransactionProcessor.class, "multiplePartitionAccesses", theEntries.size());
					throw new InvalidRequestException(Msg.code(2541) + msg);
				} else {
					retVal = retVal.mergeIds(nextRequestPartitionId);
				}
			}
```
And then in CDR, we could be shard-aware.

Reply:
I like this idea, but I think it's too ambitious for making the 7.4 release. I'd added `@Beta` and a javadoc indicating that the new API is unstable, we can clean this up in a future release.